dali/operators/image/crop/bbox_crop.cc

```diff
@@ -14,143 +14,635 @@
+      }
+      return ProspectiveCrop(true, no_crop, bounding_boxes, labels);
+    }
+
+    std::array<float, coords_size> shape, anchor, out_bounds, rel_bounds;
+    for (int i = 0; i < num_attempts_; i++) {
+      if (absolute_crop_dims) {
+        auto &crop_shape = crop_shape_[sample];
+        auto &input_shape = input_shape_[sample];
+
+        for (int d = 0; d < ndim; d++) {
+          shape[d] = static_cast<float>(crop_shape[d]);
+          out_bounds[ndim + d] = shape[d];
+          rel_bounds[ndim + d] = shape[d] / input_shape[d];
+        }
+
+        for (int d = 0; d < ndim; d++) {
+          std::uniform_int_distribution<> anchor_dist(0, input_shape[d] - crop_shape[d]);
+          anchor[d] = static_cast<float>(anchor_dist(rng));
+          out_bounds[d] = anchor[d];
+          rel_bounds[d] = anchor[d] / input_shape[d];
+        }
+      } else {  // relative dimensions
+        std::uniform_real_distribution<float> extent_dist(scale_range_.min, scale_range_.max);
+        for (int d = 0; d < ndim; d++) {
+          shape[d] = extent_dist(rng);
+          out_bounds[ndim + d] = rel_bounds[ndim + d] = shape[d];
+        }
+
+        if (!ValidAspectRatio(make_cspan(shape))) {
+          continue;
+        }
+
+        for (int d = 0; d < ndim; d++) {
+          std::uniform_real_distribution<float> anchor_dist(0.0f, 1.0f - shape[d]);
+          anchor[d] = anchor_dist(rng);
+          out_bounds[d] = rel_bounds[d] = anchor[d];
+        }
+      }
+
+      auto layout = DefaultAnchorAndShapeLayout<ndim>();  // start and shape
+      Box<ndim, float> rel_crop;
+      ReadBox(rel_crop, make_cspan(rel_bounds), layout);
+
+      Box<ndim, float> out_crop;
+      ReadBox(out_crop, make_cspan(out_bounds), layout, {});
+
+      if (!ValidOverlap(rel_crop, bounding_boxes, option.min_iou)) {
+        continue;
+      }
 
-  BoundingBoxes bounding_boxes;
-  bounding_boxes.reserve(static_cast<size_t>(boxes_tensor.dim(0)));
+      auto filtered_bboxes = bounding_boxes;
+      auto filtered_labels = labels;
+      FilterByCentroid(rel_crop, filtered_bboxes, filtered_labels);
+      if (filtered_bboxes.empty()) {
+        continue;
+      }
 
-  for (int i = 0; i < boxes_tensor.dim(0); ++i) {
-    const auto *box_data =
-        boxes_tensor.data<float>() + (i * BoundingBox::kSize);
+      for (auto &box : filtered_bboxes) {
+        box = RemapBox(box, rel_crop);
+      }
 
-    auto box = ltrb_ ? BoundingBox::FromLtrb(box_data)
-                     : BoundingBox::FromXywh(box_data);
-    bounding_boxes.emplace_back(box);
+      return ProspectiveCrop(true, out_crop, filtered_bboxes, filtered_labels);
+    }
+
+    return ProspectiveCrop();
+  }
+
+  bool ValidAspectRatio(span<const float> shape) {
+    assert(static_cast<int>(shape.size()) == coords_size);
+    for (int i = 0; i < ndim; i++) {
+      for (int j = i + 1; j < ndim; j++) {
+        if (!aspect_ratio_range_.Contains(shape[i] / shape[j]))
```

Comment:
I don't like it. It's a half-baked 3D support which will fall apart as soon as someone actually wants to use an asymmetrical aspect ratio. For real 3D, we should have _three_ aspect ratio ranges (X/Y, X/Z, Y/Z). For CT scans, for example. if you want to maintain the original aspect ratio, then X/Y would be around 1, but X/Z and Y/Z would likely be much below.

Reply:
I don't know what to do with it at this point - it would definitely complicate matters, but single aspect ratio is not very useful for real 3D data unless the user is interested in cubic sections.